libtransact/src/database/sqlite.rs

```diff
@@ -14,3 +14,778 @@
+        .map_err(|err| DatabaseError::ReaderError(format!("unable to read value: {}", err)))?;
+
+    value_iter
+        .next()
+        .transpose()
+        .map_err(|err| DatabaseError::ReaderError(format!("unable to read value: {}", err)))
+}
+
+fn execute_index_count(conn: &mut SqliteConnection, index: &str) -> Result<i64, DatabaseError> {
+    let query = format!("SELECT COUNT(index_key) FROM transact_index_{}", index);
+    let mut stmt = conn
+        .prepare_cached(&query)
+        .map_err(|err| DatabaseError::ReaderError(format!("unable to read value: {}", err)))?;
+
+    stmt.query_row(params![], |row| row.get(0))
+        .map_err(|err| DatabaseError::ReaderError(format!("unable to read value: {}", err)))
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::convert::TryInto;
+    use std::sync::atomic::{AtomicUsize, Ordering};
+
+    /// Asserts that there are COUNT many objects in DB.
+    fn assert_database_count(count: usize, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert_eq!(
+            reader.count().expect("could not count database records"),
+            count,
+        );
+    }
+
+    /// Asserts that there are are COUNT many objects in DB's INDEX.
+    fn assert_index_count(index: &str, count: usize, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert_eq!(
+            reader
+                .index_count(index)
+                .expect("unable to count index records"),
+            count,
+        );
+    }
+
+    /// Asserts that KEY is associated with VAL in DB.
+    fn assert_key_value(key: u8, val: u8, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert_eq!(
+            reader.get(&[key]).expect("unable to get value"),
+            Some(vec![val])
+        );
+    }
+
+    /// Asserts that KEY is associated with VAL in DB's INDEX.
+    fn assert_index_key_value(index: &str, key: u8, val: u8, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert_eq!(
+            reader
+                .index_get(index, &[key])
+                .expect("unable to get value by index"),
+            Some(vec![val])
+        );
+    }
+
+    /// Asserts that KEY is not in DB.
+    fn assert_not_in_database(key: u8, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert!(reader.get(&[key]).expect("unable to get value").is_none());
+    }
+
+    /// Asserts that KEY is not in DB's INDEX.
+    fn assert_not_in_index(index: &str, key: u8, db: &dyn Database) {
+        let reader = db.get_reader().expect("unable to get a database reader");
+
+        assert!(reader
+            .index_get(index, &[key])
+            .expect("unable to get value by index")
+            .is_none());
+    }
```

Comment:
Could these be put in the `database` module for reuse by other DB implementations?

Reply:
They are not particularly useful functions - I may actual considered removing them altogether.